File: .analyze_markets.py
import urllib.request, json, sys

url = 'https://api.coingecko.com/api/v3/coins/markets?vs_currency=usd&order=market_cap_desc&per_page=250&page=1&sparkline=false&price_change_percentage=1h,24h,7d'
try:
    with urllib.request.urlopen(url, timeout=30) as r:
        data = json.loads(r.read())
except Exception as e:
    print(f'ERROR: {e}', file=sys.stderr)
    sys.exit(1)

stablecoin_ids = {
    'tether','usd-coin','dai','first-digital-usd','usde','tusd','usdd','pyusd','fdusd','paxg',
    'binance-usd','frax','gemini-dollar','liquity-usd','pax-dollar','ethena-usde','paypal-usd',
    'alchemix-usd','dola-borrowing-right','usdc','crvusd','usdb','usd0','resolv-usr',
    'tether-eurt','euro-coin','stasis-eurs'
}
stablecoin_sym_prefix = ('USD','EUR','GBP','USDT','USDC','BUSD','DAI','FRAX','TUSD')
wrapped = {
    'wrapped-bitcoin','wrapped-ether','staked-ether','wrapped-steth',
    'rocket-pool-eth','coinbase-wrapped-staked-eth','lido-staked-ether'
}

filtered = []
for c in data:
    sym = c['symbol'].upper()
    cid = c['id']
    name_lower = c['name'].lower()
    if cid in stablecoin_ids or cid in wrapped:
        continue
    if any(sym == s for s in ('USDT','USDC','BUSD','DAI','FRAX','TUSD','USDP','GUSD','LUSD','USDX','SUSD','MUSD','CUSD','ZUSD','EURC','EURS','EURT')):
        continue
    if any(sym.startswith(p) for p in ('USD','EUR','GBP')):
        continue
    if 'stablecoin' in name_lower or 'pegged' in name_lower:
        continue
    vol = c.get('total_volume') or 0
    if vol < 1_000_000:
        continue
    filtered.append(c)

def fmt_price(p):
    if p is None:
        return 'N/A'
    if p >= 1000:
        return f'${p:,.0f}'
    if p >= 1:
        return f'${p:.4g}'
    if p >= 0.001:
        return f'${p:.4f}'
    return f'${p:.6f}'

def fmt_vol(v):
    if v >= 1e9:
        return f'${v/1e9:.2f}B'
    if v >= 1e6:
        return f'${v/1e6:.1f}M'
    return f'${v/1e3:.0f}K'

by_24h = sorted(filtered, key=lambda x: x.get('price_change_percentage_24h') or 0, reverse=True)

print('TOP 15 WINNERS (24h):')
for c in by_24h[:15]:
    p24 = c.get('price_change_percentage_24h') or 0
    p7 = c.get('price_change_percentage_7d_in_currency') or 0
    p1 = c.get('price_change_percentage_1h_in_currency') or 0
    vol = c.get('total_volume') or 0
    mc = c.get('market_cap') or 0
    print(f'{c["market_cap_rank"]:>4}. {c["symbol"].upper():<8} {c["name"][:24]:<25} {fmt_price(c["current_price"]):<14} 24h:{p24:+.1f}%  7d:{p7:+.1f}%  1h:{p1:+.1f}%  vol:{fmt_vol(vol)}  mc:{fmt_vol(mc)}')

print()
print('TOP 15 LOSERS (24h):')
for c in list(reversed(by_24h[-15:])):
    p24 = c.get('price_change_percentage_24h') or 0
    p7 = c.get('price_change_percentage_7d_in_currency') or 0
    p1 = c.get('price_change_percentage_1h_in_currency') or 0
    vol = c.get('total_volume') or 0
    mc = c.get('market_cap') or 0
    print(f'{c["market_cap_rank"]:>4}. {c["symbol"].upper():<8} {c["name"][:24]:<25} {fmt_price(c["current_price"]):<14} 24h:{p24:+.1f}%  7d:{p7:+.1f}%  1h:{p1:+.1f}%  vol:{fmt_vol(vol)}  mc:{fmt_vol(mc)}')

print()
top100 = [c for c in filtered if (c.get('market_cap_rank') or 999) <= 100]
top50 = [c for c in filtered if (c.get('market_cap_rank') or 999) <= 50]
pos = sum(1 for c in top100 if (c.get('price_change_percentage_24h') or 0) > 0)
vals50 = sorted([c.get('price_change_percentage_24h') or 0 for c in top50])
n = len(vals50)
median50 = (vals50[n//2-1]+vals50[n//2])/2 if n%2==0 else vals50[n//2]
print(f'MARKET PULSE: {pos}/{len(top100)} positive in top-100. Median top-50 24h: {median50:+.2f}%')

btc = next((c for c in data if c['id']=='bitcoin'), None)
eth = next((c for c in data if c['id']=='ethereum'), None)
syn = next((c for c in data if c['id']=='synapse-2'), None)
if btc:
    print(f'BTC: ${btc["current_price"]:,.0f} ({btc["price_change_percentage_24h"]:+.2f}%)')
if eth:
    print(f'ETH: ${eth["current_price"]:,.2f} ({eth["price_change_percentage_24h"]:+.2f}%)')
if syn:
    p24 = syn.get('price_change_percentage_24h') or 0
    p7 = syn.get('price_change_percentage_7d_in_currency') or 0
    vol = syn.get('total_volume') or 0
    mc = syn.get('market_cap') or 0
    print(f'SYN (check): rank={syn["market_cap_rank"]} price={fmt_price(syn["current_price"])} 24h:{p24:+.1f}% 7d:{p7:+.1f}% vol:{fmt_vol(vol)} mc:{fmt_vol(mc)}')

File: .analyze_markets.js
const fs = require('fs');
const data = JSON.parse(fs.readFileSync('.cg_markets_raw.json', 'utf8'));

const STABLE_IDS = new Set(['tether','usd-coin','dai','first-digital-usd','ethena-usde','tusd','usdd','paypal-usd','frax','usdb','true-usd','pax-gold','tether-gold','stasis-eurs','mim','alchemix-usd','dola-borrowing-right','crvusd','liquity-usd','origin-dollar','rai']);
const STABLE_SYMS = new Set(['usdt','usdc','dai','busd','tusd','usdd','usdp','gusd','frax','lusd','pyusd','fdusd','paxg','xaut','eurt','eurs','usde','crvusd','susd','usd1','usdb','musd','cusd']);
const WRAPPED = new Set(['wrapped-bitcoin','wrapped-ether','wrapped-steth','staked-ether','wrapped-eeth','coinbase-wrapped-staked-eth','rocket-pool-eth','mantle-staked-ether','binance-eth']);

function isStable(c) {
  if (STABLE_IDS.has(c.id)) return true;
  const sym = c.symbol.toLowerCase();
  if (STABLE_SYMS.has(sym)) return true;
  if (sym.startsWith('usd') || sym.startsWith('eur') || sym.startsWith('gbp')) return true;
  if (c.name.toLowerCase().includes('stablecoin')) return true;
  return false;
}

function p24(c) { return c.price_change_percentage_24h_in_currency ?? c.price_change_percentage_24h ?? 0; }
function p7d(c) { return c.price_change_percentage_7d_in_currency ?? 0; }
function p1h(c) { return c.price_change_percentage_1h_in_currency ?? 0; }

const filtered = data.filter(c => {
  if (isStable(c)) return false;
  if (WRAPPED.has(c.id)) return false;
  if ((c.total_volume || 0) < 1_000_000) return false;
  return true;
});

console.error('Filtered count:', filtered.length);

const sorted = [...filtered].sort((a, b) => p24(a) - p24(b));
const losers = sorted.slice(0, 10);
const winners = sorted.slice(-10).reverse();

function fmtPrice(p) {
  if (!p) return '?';
  if (p >= 1000) return '$' + p.toLocaleString('en', {maximumFractionDigits: 0});
  if (p >= 1) return '$' + p.toPrecision(5).replace(/\.?0+$/, '');
  if (p >= 0.01) return '$' + p.toFixed(4);
  return '$' + p.toFixed(6);
}

function fmtVol(v) {
  if (v >= 1e9) return '$' + (v/1e9).toFixed(1) + 'B';
  if (v >= 1e6) return '$' + Math.round(v/1e6) + 'M';
  return '$' + Math.round(v/1e3) + 'K';
}

console.log('\n=== WINNERS ===');
for (const c of winners) {
  const rank = c.market_cap_rank || 999;
  const tags = [];
  const p = p24(c), p7 = p7d(c), mc = c.market_cap || 0, vol = c.total_volume || 0;
  if (rank > 150 && p > 30) tags.push('[PUMP-RISK]');
  if (p > 15 && p7 > 25) tags.push('[BREAKOUT]');
  if (p > 20 && p7 < 0) tags.push('[FADE]');
  if (mc < 50_000_000) tags.push('[MICROCAP]');
  if (rank <= 20 && !tags.length) tags.push('[MAJOR]');
  console.log(c.symbol.toUpperCase() + ' (' + c.name + ') #' + rank + ' ' + fmtPrice(c.current_price) + ' 24h:' + p.toFixed(1) + '% 7d:' + p7.toFixed(1) + '% 1h:' + p1h(c).toFixed(1) + '% vol:' + fmtVol(vol) + ' mcap:' + fmtVol(mc) + ' ' + tags.join(''));
}

console.log('\n=== LOSERS ===');
for (const c of losers) {
  const rank = c.market_cap_rank || 999;
  const tags = [];
  const p = p24(c), p7 = p7d(c), mc = c.market_cap || 0, vol = c.total_volume || 0;
  if (p < -10 && mc > 0 && vol/mc > 0.25) tags.push('[CAPITULATION]');
  if (rank <= 20) tags.push('[MAJOR]');
  console.log(c.symbol.toUpperCase() + ' (' + c.name + ') #' + rank + ' ' + fmtPrice(c.current_price) + ' 24h:' + p.toFixed(1) + '% 7d:' + p7.toFixed(1) + '% 1h:' + p1h(c).toFixed(1) + '% vol:' + fmtVol(vol) + ' mcap:' + fmtVol(mc) + ' ' + tags.join(''));
}

// Market pulse
const top100 = filtered.filter(c => (c.market_cap_rank || 999) <= 100);
const green = top100.filter(c => p24(c) > 0).length;
const top50 = filtered.filter(c => (c.market_cap_rank || 999) <= 50);
const pcts50 = top50.map(c => p24(c)).sort((a,b) => a-b);
const med50 = pcts50.length ? pcts50[Math.floor(pcts50.length/2)] : 0;
const btc = data.find(c => c.id === 'bitcoin');
const eth = data.find(c => c.id === 'ethereum');

console.log('\n=== PULSE ===');
console.log('Top-100 green:', green, '/', top100.length);
console.log('Median top-50:', med50.toFixed(2) + '%');
console.log('BTC:', fmtPrice(btc && btc.current_price), (btc ? p24(btc).toFixed(1) : '?') + '%');
console.log('ETH:', fmtPrice(eth && eth.current_price), (eth ? p24(eth).toFixed(1) : '?') + '%');

// All tagged coins for Notable section
console.log('\n=== NOTABLE TAGS ===');
for (const c of sorted) {
  const rank = c.market_cap_rank || 999;
  const tags = [];
  const p = p24(c), p7 = p7d(c), mc = c.market_cap || 0, vol = c.total_volume || 0;
  if (rank > 150 && p > 30) tags.push('PUMP-RISK');
  if (p > 15 && p7 > 25) tags.push('BREAKOUT');
  if (p > 20 && p7 < 0) tags.push('FADE');
  if (p < -10 && mc > 0 && vol/mc > 0.25) tags.push('CAPITULATION');
  if (tags.length) {
    console.log(c.symbol.toUpperCase() + ' #' + rank + ' 24h:' + p.toFixed(1) + '% 7d:' + p7.toFixed(1) + '% vol/mc:' + (mc > 0 ? (vol/mc).toFixed(2) : 'N/A') + ' tags:' + tags.join(','));
  }
}

// Output all symbols for trending cross-reference
console.log('\n=== ALL IDS ===');
console.log(filtered.map(c => c.id).join(','));
